File: s3/C/tests/test_s3client.c
/*
 * Copyright Core DF — Apache License 2.0
 */

#include "../../../tests/c/testutil.h"
#include "../include/s3client.h"
#include <stdlib.h>

TEST(test_init_missing_credentials)
{
    unsetenv_safe("AWS_ACCESS_KEY_ID");
    unsetenv_safe("AWS_PROFILE");
    unsetenv_safe("S3_BUCKET");
    char *json = s3_init();
    ASSERT("init 601", json_status(json) == 601);
    free(json);
}

TEST(test_init_success)
{
    setenv("AWS_ACCESS_KEY_ID", "key", 1);
    setenv("AWS_SECRET_ACCESS_KEY", "secret", 1);
    setenv("S3_BUCKET", "my-bucket", 1);
    char *json = s3_init();
    ASSERT("init 200", json_status(json) == 200);
    free(json);
}

TEST(test_get_object_missing_bucket)
{
    unsetenv_safe("S3_BUCKET");
    char *json = s3_get_object("k", "");
    ASSERT("get 601", json_status(json) == 601);
    free(json);
}

int main(void)
{
    RUN(test_init_missing_credentials);
    RUN(test_init_success);
    RUN(test_get_object_missing_bucket);
    printf("s3 tests passed\n");
    return 0;
}

File: s3/Python/tests/test_s3client.py
"""Unit tests for s3client."""

import sys
from io import BytesIO
from pathlib import Path
from unittest.mock import MagicMock, patch

import pytest

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

import s3client


@pytest.fixture(autouse=True)
def _env(monkeypatch):
    monkeypatch.setenv("AWS_ACCESS_KEY_ID", "key")
    monkeypatch.setenv("AWS_SECRET_ACCESS_KEY", "secret")
    monkeypatch.setenv("S3_BUCKET", "my-bucket")
    monkeypatch.delenv("AWS_PROFILE", raising=False)


def test_init_missing_credentials(monkeypatch):
    monkeypatch.delenv("AWS_ACCESS_KEY_ID", raising=False)
    monkeypatch.delenv("AWS_PROFILE", raising=False)
    r = s3client.Init()
    assert r["status_code"] == 601


def test_init_success():
    assert s3client.Init() == {"status_code": 200}


@patch("s3client._client")
def test_get_object(mock_client_fn):
    client = MagicMock()
    mock_client_fn.return_value = client
    client.get_object.return_value = {"Body": BytesIO(b"payload")}

    r = s3client.GetObject("path/key.txt")

    assert r == {"status_code": 200, "content": "payload"}


@patch("s3client._client")
def test_put_object(mock_client_fn):
    client = MagicMock()
    mock_client_fn.return_value = client

    r = s3client.PutObject("k", "data")

    assert r == {"status_code": 200}
    client.put_object.assert_called_once()


@patch("s3client._client")
def test_list_objects(mock_client_fn):
    client = MagicMock()
    mock_client_fn.return_value = client
    client.list_objects_v2.return_value = {"Contents": [{"Key": "a"}, {"Key": "b"}]}

    r = s3client.ListObjects(prefix="p/")

    assert r == {"status_code": 200, "keys": ["a", "b"]}


def test_get_object_missing_bucket(monkeypatch):
    monkeypatch.delenv("S3_BUCKET", raising=False)
    r = s3client.GetObject("k", bucket="")
    assert r["status_code"] == 601
